### PR TITLE
Dynamic confirm step

### PR DESCRIPTION
**Pull request by byroot**
Fixes: https://github.com/Shopify/shipit-engine/issues/426

This is still an heavy work in progress, but I think I solved the architectural issues. I'll continue the work on this tomorrow.

@wvanbergen @dalehamel FYI.

### CODE REVIEW

File: app/models/stack.rb [hunk 1]
```diff
@@ -205,6 +205,11 @@ def checklist
     cached_deploy_spec.review_checklist.map(&:strip).select(&:present?)
   end
 
+  def checks?
+    return false unless cached_deploy_spec
```
Comment (gmalette) [inline app/models/stack.rb]: Caching the whole spec was a great idea

File: lib/shipit.rb [hunk 2]
```diff
@@ -110,8 +113,8 @@ def all_settings_present?
     ].all?(&:present?)
   end
 
-  def extra_env
-    secrets.env || {}
+  def env
+    {'SHIPIT' => '1'}.merge(secrets.env || {})
```
Comment (gmalette) [inline lib/shipit.rb]: A lot of this is actually just a great refactoring, would you mind splitting it off?

Comment (byroot) [inline lib/shipit.rb]: Indeed. I'll extract a couple PRs out of this one.

**Comment by byroot** [comment]
@gmalette @davidcornu @rafaelfranca I believe this is good to go (There is one test that fail on CI but I'll figure that out soonish).

Here's what it look like:

capture d ecran 2015-06-29 a 14 27 48

The output is streamed using the same code than for deploys, and the status is reflected by the border color: yellow: in progress, red: failed, green: success.

One thing we might need to add in a followup is a retry button.

@wvanbergen @dalehamel thoughts?

**Comment by gmalette** [comment]
Should we collapse non-stdout or something? @dalehamel for your specific case, what does the checks look like?

**Comment by byroot** [comment]
> Should we collapse non-stdout or something? 

[Buildkite's terminal](http://buildkite.github.io/terminal/) have an interesting feature: it recognize `---` and `+++` as commands to collapse the output. We might do something similar later, but I think this will do for now.

File: app/assets/javascripts/task/stream.js.coffee [hunk 3]
```diff
@@ -24,6 +24,7 @@ class @Stream
     @status = 'running'
 
   init: ({url, text, status}) ->
+    console.log(url, text, status)
```
Comment (davidcornu) [inline app/assets/javascripts/task/stream.js.coffee]: `console?.log`?

**Comment by dalehamel** [comment]
@byroot @gmalette this looks conceptually good - how many checks can we define? Each is just one command, displayed separately?

I think showing both stderr and stdout has merit. 

@gmalette our use case is that we'd like to use shipit for infrastructure (building AWS instances, etc), but we want to preview the changes before we actually apply them. This can be done in CI as well, but things might change between CI and actually shipping.

File: app/assets/javascripts/task/stream.js.coffee (hunk 3 above)
Comment (byroot) [inline app/assets/javascripts/task/stream.js.coffee]: That's a mistake. I'll :fire: it.

**Comment by byroot** [comment]
> how many checks can we define? Each is just one command, displayed separately?

As much as you want, they will be executed sequentially until one returns a non-zero status or that all have been successfully executed.

File: app/models/commit_checks.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    @status = status
+  end
+
+  def finished?
+    FINAL_STATUSES.include?(status)
+  end
+
+  def output(since: 0)
+    return fetch if since.zero?
+    redis.getrange('output', since, -1)
```
Comment (davidcornu) [inline app/models/commit_checks.rb]: How much output are we actually planning on having? Supporting `since` seems like overkill no?

**Comment by davidcornu** [comment]
```
Offenses:
app/jobs/perform_commit_checks_job.rb:2:22: C: Space inside parentheses detected.
  def perform(commit: )
                     ^
app/models/commit_checks.rb:62:38: C: Space inside parentheses detected.
  def build_commands(commands, chdir: )
```

:trollface:

File: app/models/commit_checks.rb (hunk 4 above)
Comment (byroot) [inline app/models/commit_checks.rb]: I just don't want to bork the app if a command start to output megabytes of text.

> “There's an old saying in Tennessee — I know it's in Texas, probably in Tennessee — that says, fool me once, shame on — shame on you. Fool me — you can't get fooled again.”

> -- George W Bush

File: app/assets/javascripts/task/notifications.js.coffee.erb [hunk 5]
```diff
@@ -5,12 +5,15 @@ class @Notifications
     error: '<%= image_path "deploy_error.jpg" %>'
 
   @init: (outputStream) ->
-    outputStream.addEventListener('status', new this().statusUpdated)
+    if repoName = $('[data-repo-full-name]').attr('data-repo-full-name')
+      outputStream.addEventListener('status', new this(repoName).statusUpdated)
```
Comment (davidcornu) [inline app/assets/javascripts/task/notifications.js.coffee.erb]: `new this` looks odd to me but I guess that's the only way of supporting inheritance.

File: app/assets/javascripts/task/notifications.js.coffee.erb [hunk 6]
```diff
@@ -5,12 +5,15 @@ class @Notifications
...
+
+  constructor: (@repoName) ->
 
   statusUpdated: (task) =>
     return if task.status in ['pending', 'running']
     return unless $.notifyCheck() == $.NOTIFY_ALLOWED
-    $.notify(@url(task), @title(task), @message(task))
+    $.notify(@url(task), repoName, @message(task))
```
Comment (davidcornu) [inline app/assets/javascripts/task/notifications.js.coffee.erb]: `@repoName`?

File: app/models/commit_checks.rb [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    commands = StackCommands.new(stack)
+    commands.with_temporary_working_directory(commit: commit) do |directory|
+      deploy_spec = DeploySpec::FileSystem.new(directory, stack.environment)
+      Bundler.with_clean_env do
+        capture_all(build_commands(deploy_spec.dependencies_steps, chdir: directory))
+        capture_all(build_commands(deploy_spec.review_checks, chdir: directory))
+      end
+    end
+  rescue Command::Error
+    self.status = 'failed'
```
Comment (davidcornu) [inline app/models/commit_checks.rb]: Does this get logged anywhere?

Comment (davidcornu) [inline app/models/commit_checks.rb]: I guess `capture_all` would give you debug info

Comment (byroot) [inline app/models/commit_checks.rb]: Yeah the status is so the client side knows it's over. The command status etc is appended to the output.

File: app/models/commit_checks.rb [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+  rescue
+    self.status = 'error'
+    raise
+  else
+    self.status = 'success'
```
Comment (davidcornu) [inline app/models/commit_checks.rb]: Any reason not to put this after line `18`?

File: app/models/commit_checks.rb [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+  end
+
+  def fetch
+    @output ||= begin
```
Comment (davidcornu) [inline app/models/commit_checks.rb]: Having both a `@output` instance variable that is populated by `fetch` and a `output` method that does something completely different seems odd to me.

**Comment by wvanbergen** [comment]
@byroot Does displaying this page create a deploy lock? Or is is potentially possible to have a bunch of people requesting this page at the same time, and/or somebody else deploying when showing this page?

File: app/models/commit_checks.rb (hunk 8 above)
Comment (byroot) [inline app/models/commit_checks.rb]: Not really no.

File: app/models/commit_checks.rb [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    redis.set('status', status, ex: OUTPUT_TTL)
+    @status = status
+  end
+
+  def finished?
+    FINAL_STATUSES.include?(status)
+  end
+
+  def output(since: 0)
+    return fetch if since.zero?
```
Comment (davidcornu) [inline app/models/commit_checks.rb]: Doesn't this mean you can call `output` and get nothing back if `since > 0` and the output hasn't been populated yet?

**Comment by byroot** [comment]
> Does displaying this page create a deploy lock? Or is is potentially possible to have a bunch of people requesting this page at the same time, and/or somebody else deploying when showing this page?

@wvanbergen yes this page is totally independent from a deploy. The commands are executed in a temporary clone of the stack.

If 2 people are visualizing this page for the same commit at the same, they'll share the output, but that's all.

File: app/models/commit_checks.rb (hunk 10 above)
Comment (byroot) [inline app/models/commit_checks.rb]: Indeed.

File: app/models/commit_checks.rb (hunk 4 above)
Comment (davidcornu) [inline app/models/commit_checks.rb]: Sure

File: app/models/commit_checks.rb (hunk 10 above)
Comment (byroot) [inline app/models/commit_checks.rb]: I'll see if I can improve this class encapsulation.

Comment (davidcornu) [inline app/models/commit_checks.rb]: Shouldn't we instead check of the command has been run?

File: app/models/commit_checks.rb (hunk 8 above)
Comment (davidcornu) [inline app/models/commit_checks.rb]: Seems odd to put the success case after all the error cases as a `rescue/else`.

**Comment by davidcornu** [comment]
re: rubocop

I prefer having a trailing space after required keyword arguments but that might just be me

```ruby
def my_method(required_arg: )
                           ^
```

**Comment by davidcornu** [comment]
Some minor concerns around lazy-loading the command output but overall LGTM.

**Comment by gmalette** [comment]
LGTM. I didn't :tophat: but I saw this was using `OutputStream` which means it'll have colorization, right?

**Comment by byroot** [comment]
> which means it'll have colorization, right?

Indeed.

**Comment by byroot** [comment]
@davidcornu 5a7b782 refactored `CommitChecks`. I think you'll like it better.

I'll merge on green CI.

One last thing, as of now these checks only execute for deploys, meaning not for rollbacks. Not sure if they should as you don't want to delay rollbacks too much. But if you think they should let me know I can add them.

**Comment by dalehamel** [comment]
:heart: :sparkles: :heart: 

:clap: 

:gift: 

:tada:

File: app/models/commit_checks.rb (hunk 8 above)
Comment (davidcornu) [inline app/models/commit_checks.rb]: I still don't like this

**Comment by davidcornu** [comment]
> @davidcornu 5a7b782 refactored `CommitChecks`. I think you'll like it better.

:+1:

File: app/models/commit_checks.rb (hunk 8 above)
Comment (byroot) [inline app/models/commit_checks.rb]: What the `else`? It's better than putting it inside the main block, that way this statement can't be rescued by the block.

Comment (davidcornu) [inline app/models/commit_checks.rb]: > that way this statement can't be rescued by the block.

What does that add though? It's just an `attr_accessor`, if that breaks I'd be a lot more worried.

Comment (byroot) [inline app/models/commit_checks.rb]: > It's just an `attr_accessor`

It's not :wink:

Comment (davidcornu) [inline app/models/commit_checks.rb]: Ah. My bad

**Pull request merged**